File: 백준/Gold/10775. 공항/공항.py
import sys

input = sys.stdin.readline

G = int(input())
P = int(input())

airplanes =  [int(input()) for _ in range(P)]
parents = [i for i in range(G + 1)]


def find_parent(x):
    global parents

    if parents[x] == x:
        return x

    parents[x] = find_parent(parents[x])

    return parents[x]


def union(x, y):
    global parents

    px, py = find_parent(x), find_parent(y)
    if px == py:
        return

    parents[px] = py


answer = 0
for airplane in airplanes:
    parent = find_parent(airplane)
    
    if parent == 0:
        break
    union(parent, parent - 1)
    answer += 1
print(answer)
    

File: 백준/Gold/10775. 공항/공항.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class Main{

    private static int n, m;
    private static int[] parents;
    private static int[] gates;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        m = Integer.parseInt(br.readLine());

        parents = new int[n + 1];
        gates = new int[m];

        for(int i=1 ; i<=n ; i++){
            parents[i] = i;
        }

        for(int i=0 ; i<m ; i++){
            gates[i] = Integer.parseInt(br.readLine());
        }

        int answer = 0;
        for(int i=0 ; i<m ; i++){
            int parent = findParent(gates[i]);

            if(parent == 0)
                break;

            union(parent, parent - 1);
            
            answer++;
        }

        System.out.println(answer);

    }

    private static int findParent(int x){
        if(x == parents[x]){
            return x;
        }

        return parents[x] = findParent(parents[x]);
    }

    private static void union(int x, int y){
        int px = findParent(x), py = findParent(y);

        if(px == py){
            return;
        }

        parents[px] = py;
    }
}